protocols/mdns/src/behaviour.rs

```diff
@@ -379,6 +380,7 @@ impl fmt::Debug for DiscoveredAddrsIter {
 }
 
 /// Iterator that produces the list of addresses that have expired.
+#[derive(Clone)]
```

Comment:
Why do the iterators need to be `Clone` too?

Reply:
As the compiler suggests. Because the type is not a alias but a struct holding the iterator. All members should also be `Clone` when deriving `Clone` and a struct is not `Clone` by default.

Reply:
> All members should also be `Clone` when deriving `Clone` and a struct is not `Clone` by default.

Yes but why do these two structs need to be `Clone` in the first place? Is it not sufficient that the event they return is `Clone`? This means you can collect the iterator into a `Vec<mdns::Event>` and clone that vector as often as you want.

Iterators in Rust are typically single-use, I am not sure I want to make an API commitment that these ones can be cloned.

Reply:
I'm not comfortable with cloning iterators, too. So originally I changed the enum to hold `SmallVec` making it a breaking change. I'm experimenting with event listeners and type information makes delegating events to appropriate places easier. But if the event is single-use, it can't be send to multiple places with the type it comes from, causing inconsistency. I'm not so comfortable with that either.

Reply:
I don't think it's a good idea to construct data structures in places that shouldn't.

Reply:
Plus, the two structs have private field and there's no constructor for both of them currently, so there's actually no way to construct them outside the crate, let alone constructing `mdns::Event`.

Reply:
I am not sure I understand what you are getting at. With `mdns::Event` now being `Clone`, you can do the following with the iterator:

```rust
let peers = iter.collect::<Vec<_>>();
let peers2 = peers.clone();
```

What I am suggesting is to remove the `derive(Clone)` from the two iterator structs.

Reply:
As compiler suggests, you cannot derive `Clone` for `mdns::Event` unless the two structs are also `Clone`, because the variants hold them, and cloning the variant also clone the struct it holds. You can of course collect the results from the iterator with type `Vec<(PeerId,Multiaddr)>`, but you cannot put it back into the `Event` it comes with without access to the private field the struct holds. Although I can very well be designing a type as I want and implement `From<T>` and `Into<T>`, I still think it's a good idea to keep the original type information in case of confusion. It's also easier to trace where it comes from. If cloning in this way cannot be achieved, I'm reverting this PR. Everything started with the **unnecessary** iterator implementation, without which everyone is happy.

Reply:
Oh my god, I am a moron. Sorry for all this back and forth, I didn't look at the code properly. I thought `mdns::Event` was within those iterators and not the event holding the iterators :man_facepalming: 

> but you cannot put it back into the `Event`

That is true. Can you achieve your usecase without putting it back in? Typically, these events are only consumed by applications and translated into whatever data structure they need.

Sorry for going back on this but after having looked at everything properly now, I think having it `Clone` is unnecessary because you can simply extract the data and clone that.

Reply:
Again, collecting result is totally fine. But the type information is lost, making it harder to trace where it comes from. At least we can try to preserve the information for as long as possible before conversion is absolutely needed, and that's the whole point.

Reply:
> Again, collecting result is totally fine. But the type information is lost, making it harder to trace where it comes from. At least we can try to preserve the information for as long as possible before conversion is absolutely needed, and that's the whole point.

I guess that is fair. Happy to ship this. Thank you for your patience :)

Reply:
> > but you cannot put it back into the `Event`
> 
> That is true. Can you achieve your usecase without putting it back in? Typically, these events are only consumed by applications and translated into whatever data structure they need.

I am making an application and I can tell you very confidently that there's still some way to go before it gets converted. Plus, if all I want is just a vector, I need to collect the result manually, which is way more baffling than type `iter()` or `into_iter()` to get an iterator. I'll get upset in the former case while complete fine in the latter case because it's normal, it's part of life.